crypto/sig/README.md

```diff
@@ -0,0 +1,85 @@
+adversary. This caused many high profile security breaks, when the PRNG used for the nonces
+didn't produce truly random nonces.
+
+In RFC6979, ed25519, and much later work, the idea was that the signature should be deterministic,
+and the nonce should be pseudorandomly generated from the message, and entropy connected to the
+private key. This ensures that only one nonce is ever used in connection to a particular message.
+If a PRF is used to compute the nonce, then the nonce is hard to distinguish from random even if
+the messages that are signed are adversarially chosen.
+
+In the `sign` function in this crate, we take as an assumption that secret-prefix-Blake2b is a PRF.
+
+The ed25519 manuscript from 2011-09-26 has remarks in the section "pseudorandom generation of r", where
+`r` is the nonce, which support this idea (http://ed25519.cr.yp.to/ed25519-20110926.pdf):
+
+> This idea of generating random signatures in a secretly deterministic way, in particular obtaining
+> pseudorandomness by hashing a long-term secret key together with the input message, was proposed by
+> Barwood in [9]; independently by Wigley in [79]; a few months later in a patent application [57]
+> by Naccache, M’Ra ̈ıhi, and  Levy-dit-Vehel; later  by M’Ra ̈ıhi, Naccache, Pointcheval, and Vaudenay
+> in [55]; and much later by Katz and Wang in [47]. The patent application was abandoned in 2003.
+>
+> Standard PRF hypotheses imply that this pseudorandom session key `r` is indistinguishable from a
+> truly random string generated independently for each `M`, so there is no loss of security.
+> Well-known length-extension properties prevent secret-prefix SHA-512 from being a PRF, but also do
+> not threaten the securityof Ed25519-SHA-512, since `r` is not visible to the attacker. All remaining
+> SHA-3 candidates are explicitly designed to be PRFs, and we will not hesitate to recommend
+> `Ed25519-SHA-3` after SHA-3 is standardized. It would of course also be safe to generate `r` with
+> a cipher such as AES, combined with standard PRF-stretching mechanisms to support a long input;
+> but we prefer to reuse `H` to save area in hardware implementations.
+
+We point out that Blake was one of the SHA-3 finalists, and was also explicitly designed to model a PRF.
+We prefer Blake2b here because it reduces the number of different hash functions in our system overall.
+
+Assuming the Blake2b has the secret-prefix-PRF property, we can say that the signatures created this way
+are hard to distinguish from signatures created where the nonce is truly uniformly random, even if the messages
+that are signed are adversarially chosen. So, if Schnorrkel is secure when the nonces are truly random and
+the RNG is the OS-RNG, or, when the nonce is created using the mini-secret-key expansion, then this should also be secure.
+
+Rng required by Schnorrkel
+--------------------------
+
+Schnorrkel API deviates from RFC6979 and ed25519 in requiring the signer to provide a CSPRNG, which they
```

Comment:
I think the Schnorrkel API allows us to bypass the OS rng via `attach_rng`, but I haven't been able to devote time to trying this out yet.

Reply:
yeah but my point is, RFC6979 and ed25519 don't require an RNG at all. like, that's what deterministic signatures means. i think Schnorrkel has departed from earlier work in asking us for an RNG at all.